src/solvers/flattening/bv_pointers.cpp

```diff
@@ -172,8 +172,38 @@ literalt bv_pointerst::convert_rest(const exprt &expr)
       const bvt &bv0=convert_bv(operands[0]);
       const bvt &bv1=convert_bv(operands[1]);
 
-      return bv_utils.rel(
-        bv0, expr.id(), bv1, bv_utilst::representationt::UNSIGNED);
+      const pointer_typet &type0 = to_pointer_type(operands[0].type());
+      bvt offset_bv0 = offset_literals(bv0, type0);
+
+      const pointer_typet &type1 = to_pointer_type(operands[1].type());
+      bvt offset_bv1 = offset_literals(bv1, type1);
+
+      // comparison over pointers to distinct objects is undefined and yields a
+      // non-deterministic result; alternatively, we could do a comparison over
+      // the integer representation of a pointer
+
+      // do the same-object-test via an expression as this may permit re-using
+      // already cached encodings of the equality test
+      const exprt same_object = ::same_object(operands[0], operands[1]);
+      const bvt &same_object_bv = convert_bv(same_object);
+      CHECK_RETURN(same_object_bv.size() == 1);
+
+      literalt l = prop.new_variable();
+
+      if(!same_object_bv[0].is_false())
```

Comment:
I like this optimisation but if we are being super strict, shouldn't the new literal be inside the if too.

Reply:
No, I don't think so: the encoding is such that the objects not matching leaves it to the solver to choose a value for that literal. This also means that `__CPROVER_assume( p < &array + LENGTH );` does not constrain `p` to be in the same object. I'm expanding the regression test to include this and demonstrate that it isn't being constrained.

Reply:
Additional note: if you _want_ to constrain it to the same object, this can be done using `__CPROVER_pointer_object(p) == __CPROVER_pointer_object(&array)`.

Reply:
Thinking about it; is this correct?  The case in which the if is not taken is when it can be (encoding time) determined that the pointers cannot point to the same object, right?  In which case ISO 9899 says that the comparison is undefined behaviour.  With `goto_check` we should be able to detect this case and users shouldn't rely on something that happens after undefined behaviour.  So we are back to the question of what we do after UB?  Your solution is to make it non-deterministic, but it would be no less correct for it to be false or true.

I think non-deterministic probably is the best trade off but A. I want to make sure that I understand what is going on and B. can you add a comment to saying that it is non-deterministic, the case and why.

Reply:
We do correctly cover this case in `goto_check`, reporting "same object violation in ..." (tried on the regression test I'm including here). So I think we're safe on that side. The question is what the back-end should do in such cases. As you say, and behaviour is permitted.

Reply:
I wonder if there is a doxygen-based way of commenting in the source and building a page / list of ... definitions of undefined behaviour, i.e. what we do.

Reply:
We could introduce a new group `\defgroup undef Undefined Behaviour` and then do `\addtogroup undef`.

Reply:
I would use different terminology here. "Undefined behavior" has specific ramifications when dealing with C code (despite the name), and in a tool for checking C this is what a reader will think of.

There are pros and cons to using nondeterminism for such cases. As starting point, take a look at the discussion around the meaning of division by zero in SMT-LIB.

The choice 'nondeterminism' is safe when doing satisfiability checking. It will start to bite once quantifiers are used.

Reply:
So if we make a choice: should it be always true or always false? I'd intuitively pick the latter, but I'm not sure I have arguments supporting this.

Reply:
I'd also go for 'false'; this is what IEEE did with NaNs.

Reply:
BTW, doing that requires a review of the simplifier. In particular, the rule !(a>b) = a<=b no longer holds.

Reply:
If you wanted to retain that rule, and not introduce an implicit existential quantification, you could go back to the total ordering that was there before.

Reply:
I have now implemented "false" as the choice for comparison of pointers to distinct objects, and have also updated the simplifier. Thank you for calling this out, @kroening!